src/storage/sql_table.cpp

```diff
@@ -250,11 +250,18 @@ std::pair<bool, storage::TupleSlot> SqlTable::Update(transaction::TransactionCon
     ProjectedRow *new_pr = new_pair.first.InitializeRow(new_buffer);
     StorageUtil::CopyProjectionIntoProjection(*old_pr, old_pair.second, tables_.at(old_version).layout, new_pr,
                                               new_pair.second);
-    // 3. Insert the row into new table
-    storage::TupleSlot new_slot = tables_.at(version_num).data_table->Insert(txn, *new_pr);
 
-    // 4. Delete the old row
-    tables_.at(old_version).data_table->Delete(txn, slot);
+    // 3. Delete the old row
+    bool succ = tables_.at(old_version).data_table->Delete(txn, slot);
+
+    // 4. Insert the row into new table
+    storage::TupleSlot new_slot;
+    if (succ) {
+      new_slot = tables_.at(version_num).data_table->Insert(txn, *new_pr);
+    } else {
+      // someone else deleted the old row, the tupleslot is not valid anymore
+      return {false, slot};
+    }
 
     // 5. Update the new row
     tables_.at(version_num).data_table->Update(txn, new_slot, redo);
```

Comment:
Why are we calling `Update`?  We should be able to merge the changes into the `ProjectedRow` we are about to `Insert` and save a round-trip.

Reply:
Looks good pending final Travis checks for formatting checks.